src/Rematch.jl

```diff
@@ -112,11 +123,32 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
             end
         end
     elseif @capture(pattern, T_(subpatterns__))
-        @assert occursin(r"^[A-Z]", string(T)) "Pattern $pattern looks like a struct pattern but $T is probably not a struct type."
         # struct
-        push!(asserts, quote
-            assert_num_fields($(esc(T)), $(length(subpatterns)))
-        end)
+        @assert occursin(r"^[A-Z]", string(T)) "Pattern $pattern looks like a struct pattern but $T is probably not a struct type."
+        len = length(subpatterns)
+        named_fields = [pat.args[1] for pat in subpatterns if (pat isa Expr) && pat.head == :kw]
+        named_count = length(named_fields)
+        @assert named_count == length(unique(named_fields)) "Pattern $pattern has duplicate named arguments: $(named_fields)"
+        @assert named_count == 0 || len == named_count "Pattern $pattern mixes named and positional arguments"
+        if named_count == 0
+            # Uses positional variables
```

Comment:
I'm not really sure what this comment means. Does "positional variables" refer to a general metaprogramming concept that I'm missing?

Can you expand that to better explain what you meant by that and why the comment was needed?
I think what you're trying to get at is something like that you need to create these new names, because they need to be unique to this type `T`, not just sheltered from user code in the normal macro way?

Reply:
Maybe it should be "positional/named arguments" instead of "variables". Given a `function foo(x=0, y=true)` with keyword arguments (maybe i need a ; in there), then:
* `foo(1, false)` uses positional variables/arguments
* `foo(y=false, x=1)` uses named variables/arguments

The comment is to explain what branch we're in.

The gensym'ed names are to avoid clashes with the rest of the program, the type `T` is only included to aid debugging. I'll switch to `local` declarations.

Reply:
> The gensym'ed names are to avoid clashes with the rest of the program, the type T is only included to aid debugging. I'll switch to local declarations.

 Ah yeah, duh, you're right. thanks. I'll close that thread.

Reply:
> The comment is to explain what branch we're in.

Oh gotcha thanks. yeah that wasn't clear. if you could just expand it a bit i think that would help!

>  "positional/named arguments" instead of "variables".

+1
Maybe:
```
# Pattern uses positional arguments to refer to fields
```
or something?